File: src/TmrOverlay.App/Runtime/BuildFreshnessChecker.cs
namespace TmrOverlay.App.Runtime;

internal static class BuildFreshnessChecker
{
    private const string SolutionFileName = "tmrOverlay.sln";
    private static readonly string[] SourceSearchPatterns =
    [
        "*.cs",
        "*.csproj",
        "appsettings.json",
        "Package.swift",
        "*.swift",
        "*.md"
    ];

    public static BuildFreshnessResult Check()
    {
        var repositoryRoot = FindRepositoryRoot(AppContext.BaseDirectory);
        var buildStamp = FindBuildStamp(AppContext.BaseDirectory);

        if (repositoryRoot is null || buildStamp is null)
        {
            return BuildFreshnessResult.Current;
        }

        var buildTimeUtc = buildStamp.LastWriteTimeUtc;
        var newestSource = FindNewestSource(repositoryRoot);

        if (newestSource is null || newestSource.LastWriteTimeUtc <= buildTimeUtc.AddSeconds(2))
        {
            return BuildFreshnessResult.Current;
        }

        var relativePath = Path.GetRelativePath(repositoryRoot, newestSource.FullName);
        return new BuildFreshnessResult(
            SourceNewerThanBuild: true,
            Message: $"Local source is newer than this build; rebuild recommended. Newest: {relativePath}");
    }

    private static FileSystemInfo? FindBuildStamp(string appBaseDirectory)
    {
        if (string.IsNullOrWhiteSpace(appBaseDirectory) || !Directory.Exists(appBaseDirectory))
        {
            return null;
        }

        var appDirectory = new DirectoryInfo(appBaseDirectory);
        foreach (var fileName in new[] { "TmrOverlay.App.exe", "TmrOverlay.App.dll" })
        {
            var candidate = Path.Combine(appDirectory.FullName, fileName);
            if (File.Exists(candidate))
            {
                return new FileInfo(candidate);
            }
        }

        return appDirectory;
    }

    private static FileInfo? FindNewestSource(string repositoryRoot)
    {
        var roots = new[]
        {
            Path.Combine(repositoryRoot, "src"),
            Path.Combine(repositoryRoot, "tests"),
            Path.Combine(repositoryRoot, "local-mac", "TmrOverlayMac"),
            Path.Combine(repositoryRoot, "README.md"),
            Path.Combine(repositoryRoot, "VERSION.md")
        };

        return roots
            .SelectMany(EnumerateSourceFiles)
            .OrderByDescending(file => file.LastWriteTimeUtc)
            .FirstOrDefault();
    }

    private static IEnumerable<FileInfo> EnumerateSourceFiles(string path)
    {
        if (File.Exists(path))
        {
            yield return new FileInfo(path);
            yield break;
        }

        if (!Directory.Exists(path))
        {
            yield break;
        }

        foreach (var pattern in SourceSearchPatterns)
        {
            foreach (var file in Directory.EnumerateFiles(path, pattern, SearchOption.AllDirectories))
            {
                if (file.Contains($"{Path.DirectorySeparatorChar}bin{Path.DirectorySeparatorChar}", StringComparison.OrdinalIgnoreCase)
                    || file.Contains($"{Path.DirectorySeparatorChar}obj{Path.DirectorySeparatorChar}", StringComparison.OrdinalIgnoreCase)
                    || file.Contains($"{Path.DirectorySeparatorChar}.build{Path.DirectorySeparatorChar}", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                yield return new FileInfo(file);
            }
        }
    }

    private static string? FindRepositoryRoot(string startingDirectory)
    {
        var directory = new DirectoryInfo(startingDirectory);

        while (directory is not null)
        {
            var solutionPath = Path.Combine(directory.FullName, SolutionFileName);
            if (File.Exists(solutionPath))
            {
                return directory.FullName;
            }

            directory = directory.Parent;
        }

        return null;
    }
}

internal sealed record BuildFreshnessResult(bool SourceNewerThanBuild, string? Message)
{
    public static BuildFreshnessResult Current { get; } = new(SourceNewerThanBuild: false, Message: null);
}

File: local-mac/TmrOverlayMac/Sources/TmrOverlayMac/Runtime/BuildFreshnessChecker.swift
import Foundation

struct BuildFreshnessResult {
    let sourceNewerThanBuild: Bool
    let message: String?

    static let current = BuildFreshnessResult(sourceNewerThanBuild: false, message: nil)
}

enum BuildFreshnessChecker {
    static func check() -> BuildFreshnessResult {
        guard let repositoryRoot = findRepositoryRoot(),
              let executableURL = Bundle.main.executableURL ?? executableURLFromArguments() else {
            return .current
        }

        let executableDate = modificationDate(executableURL)
        guard let newestSource = newestSourceFile(in: repositoryRoot),
              modificationDate(newestSource) > executableDate.addingTimeInterval(2) else {
            return .current
        }

        let relativePath = newestSource.path.replacingOccurrences(of: repositoryRoot.path + "/", with: "")
        return BuildFreshnessResult(
            sourceNewerThanBuild: true,
            message: "Local source is newer than this build; rebuild recommended. Newest: \(relativePath)"
        )
    }

    private static func executableURLFromArguments() -> URL? {
        guard let executablePath = CommandLine.arguments.first else {
            return nil
        }

        return URL(fileURLWithPath: executablePath).standardizedFileURL
    }

    private static func newestSourceFile(in repositoryRoot: URL) -> URL? {
        let roots = [
            repositoryRoot.appendingPathComponent("local-mac/TmrOverlayMac/Package.swift"),
            repositoryRoot.appendingPathComponent("local-mac/TmrOverlayMac/Sources", isDirectory: true),
            repositoryRoot.appendingPathComponent("local-mac/TmrOverlayMac/Tests", isDirectory: true),
            repositoryRoot.appendingPathComponent("src", isDirectory: true),
            repositoryRoot.appendingPathComponent("README.md")
        ]

        return roots
            .flatMap(sourceFiles)
            .max { modificationDate($0) < modificationDate($1) }
    }

    private static func sourceFiles(in url: URL) -> [URL] {
        if FileManager.default.fileExists(atPath: url.path), !isDirectory(url) {
            return [url]
        }

        guard isDirectory(url),
              let enumerator = FileManager.default.enumerator(
                at: url,
                includingPropertiesForKeys: [.isRegularFileKey, .contentModificationDateKey],
                options: [.skipsHiddenFiles]
              ) else {
            return []
        }

        return enumerator
            .compactMap { $0 as? URL }
            .filter { isSourceFile($0) && !isBuildOutput($0) }
    }

    private static func isSourceFile(_ url: URL) -> Bool {
        ["swift", "cs", "csproj", "json", "md"].contains(url.pathExtension.lowercased())
    }

    private static func isBuildOutput(_ url: URL) -> Bool {
        let path = url.path
        return path.contains("/.build/")
            || path.contains("/bin/")
            || path.contains("/obj/")
    }

    private static func isDirectory(_ url: URL) -> Bool {
        ((try? url.resourceValues(forKeys: [.isDirectoryKey]))?.isDirectory) == true
    }

    private static func modificationDate(_ url: URL) -> Date {
        ((try? url.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate) ?? .distantPast
    }

    private static func findRepositoryRoot() -> URL? {
        var directory = URL(fileURLWithPath: FileManager.default.currentDirectoryPath).standardizedFileURL

        while true {
            if FileManager.default.fileExists(atPath: directory.appendingPathComponent("tmrOverlay.sln").path) {
                return directory
            }

            let parent = directory.deletingLastPathComponent()
            if parent.path == directory.path {
                return nil
            }

            directory = parent
        }
    }
}
